README.md

# RPC

![PyPI - Python Version](https://img.shields.io/pypi/pyversions/twentyc.rpc)
[![PyPI version](https://badge.fury.io/py/twentyc.rpc.svg)](https://badge.fury.io/py/twentyc.rpc)
[![Build Status](https://travis-ci.org/20c/twentyc.rpc.svg?branch=master)](https://travis-ci.org/20c/twentyc.rpc)
[![Codecov](https://img.shields.io/codecov/c/github/20c/twentyc.rpc/master.svg?maxAge=2592000)](https://codecov.io/github/20c/twentyc.rpc)
[![Language grade: Python](https://img.shields.io/lgtm/grade/python/g/20c/twentyc.rpc.svg?logo=lgtm&logoWidth=18)](https://lgtm.com/projects/g/20c/twentyc.rpc/context:python)


client for 20c's RESTful API



CHANGELOG.md
# Changelog


## Unreleased


## 1.0.0
### Added
- linting
- moved to github workflows
### Removed
- python 2 support


## 0.4.0
### Added
- py3 compatibility
### Changed
- updated test requirements


## 0.3.5
### Changed
- updated requirements

tests/test_client.py
"""
Unittests for twentyc.rpc.client
"""


import json
import uuid

import pytest

import twentyc.rpc.client as client

## PREP

INVALID_DATA_ERROR = {"name": "value required"}


def obj(id, name=None):
    return {"_id": id, "name": (name or str(uuid.uuid4()))}


def objs(num):
    return [obj(i) for i in range(num)]


def response(num, status_code=200, meta={}, headers={}):
    return {
        "status_code": status_code,
        "content": json.dumps({"meta": meta, "data": objs(num)}),
        "headers": headers,
    }


testData = {
    "_throw": json.dumps({"meta": {"error": "TEST_ERROR"}}),
    "_load": json.dumps({"meta": {}, "data": objs(1)}),
    "GET": {
        "/api/obj": response(10),
        "/api/obj/1": response(1),
        "/api/obj/2": response(0, status_code=404),
    },
    "POST": {
        "/api/obj_201": response(
            0, status_code=201, headers={"location": "/api/obj/1"}
        ),
        "/api/obj_401": response(0, status_code=401),
        "/api/obj_400": response(0, status_code=400, meta=INVALID_DATA_ERROR),
    },
    "PUT": {
        "/api/obj_200/1": response(1),
        "/api/obj_401/1": response(0, status_code=401),
        "/api/obj_400/1": response(0, status_code=400, meta=INVALID_DATA_ERROR),
    },
}


class DummyResponse:
    def __init__(self, status_code, content, headers={}):
        self.status_code = status_code
        self.content = content
        self.headers = headers

    @property
    def data(self):
        return json.loads(self.content)

    def read(self, *args, **kwargs):
        return self.content

    def getheader(self, name):
        return self.headers.get(name)

    def json(self):
        return self.data


class DummyClient(client.RestClient):

    testData = testData

    def __init__(self, **kwargs):
        super().__init__("http://localhost", **kwargs)

    def _request(self, typ, id=0, method="GET", data=None, params=None, url=None):

        """
        Instead of requesting to a HTTP connection we retrieve pre-crafted
        responses from testData
        """

        print("URL", url, typ, id)

        if not url:
            url = "/api/%s" % typ
            if id:
                url = f"{url}/{id}"
        else:
            url = url.replace("http://localhost", "")

        self._response = DummyResponse(**(self.testData.get(method).get(url)))
        return self._response


## TESTS


def test_instantiate_default():

    client = DummyClient()

    assert client.url == "http://localhost"
    assert client.user == None
    assert client.password == None
    assert client.timeout == None
    assert client.verbose == False


def test_instantiate_arguments():

    kwargs = {
        "user": "user",
        "password": "pass",
        "timeout": 4000,
        "verbose": True,
    }

    client = DummyClient(**kwargs)
    for k, v in list(kwargs.items()):
        assert getattr(client, k) == v


def test_url_update():
    client = DummyClient()
    assert client.url_update(path="/apu") == "http://localhost/apu"


def test__throw():
    response_404 = DummyResponse(404, testData.get("_throw"))
    response_401 = DummyResponse(401, testData.get("_throw"))
    response_400 = DummyResponse(400, testData.get("_throw"))
    response_500 = DummyResponse(500, "{}")

    c = DummyClient()

    with pytest.raises(client.NotFoundException) as exc:
        c._throw(response_404, response_404.data)

        assert exc.value.message == "404 TEST_ERROR"

    with pytest.raises(client.PermissionDeniedException) as exc:
        c._throw(response_401, response_401.data)

        assert exc.value.message == "401 TEST_ERROR"

    with pytest.raises(client.InvalidRequestException) as exc:
        c._throw(response_400, response_400.data)

        assert exc.value.message == "400 TEST_ERROR"

    with pytest.raises(LookupError) as exc:
        c._throw(response_404, response_404.data)

    with pytest.raises(IOError) as exc:
        c._throw(response_401, response_401.data)

    with pytest.raises(ValueError) as exc:
        c._throw(response_400, response_400.data)

    with pytest.raises(Exception) as exc:
        c._throw(response_500, response_500.data)

        assert exc.value.message == "500 Internal error: Unknown"


def test__load():

    response_200 = DummyResponse(200, testData.get("_load"))
    response_404 = DummyResponse(404, testData.get("_throw"))

    c = DummyClient()

    data = c._load(response_200)
    assert type(data) == list
    assert len(data) == 1
    assert data[0] == response_200.data.get("data")[0]

    with pytest.raises(client.NotFoundException) as exc:
        c._load(response_404)


def test__mangle_data():

    before = {"pk": 1, "_rev": 123, "_id": 1}

    after = {"id": 1}

    c = DummyClient()

    c._mangle_data(before)
    assert before == after


def test_get():

    c = DummyClient()
    data = c.get("obj", 1)
    assert data == c._response.data.get("data")

    with pytest.raises(client.NotFoundException) as exc:
        c.get("obj", 2)


def test_all():

    c = DummyClient()
    data = c.all("obj")
    assert len(data) == 10
    assert data == c._response.data.get("data")


def test_create():

    c = DummyClient()

    expected = json.loads(testData.get("GET").get("/api/obj/1").get("content"))
    data = c.create("obj_201", expected.get("data")[0])
    assert data == expected.get("data")

    with pytest.raises(client.PermissionDeniedException) as exc:
        data = c.create("obj_401", expected.get("data")[0])

    with pytest.raises(client.InvalidRequestException) as exc:
        data = c.create("obj_400", expected.get("data")[0])
        assert exc.value.extra == INVALID_DATA_ERROR


def test_update():

    c = DummyClient()

    expected = json.loads(testData.get("PUT").get("/api/obj_200/1").get("content"))

    data = c.update("obj_200", 1, **expected.get("data")[0])
    assert data == expected.get("data")


src/twentyc/rpc/client.py
#!/bin/env python


import requests
from urllib import parse


class NotFoundException(LookupError):
    pass


class PermissionDeniedException(IOError):
    pass


class InvalidRequestException(ValueError):
    def __init__(self, msg, extra):
        super().__init__(self, msg)
        self.extra = extra


class RestClient:
    def __init__(self, url, **kwargs):
        """
        RESTful client
        """
        self.url = url
        self._url = parse.urlparse(self.url)
        self.user = None
        self.password = None
        self.timeout = None
        self.verbose = False

        # overwrite any param from keyword args
        for k in kwargs:
            if hasattr(self, k):
                setattr(self, k, kwargs[k])

    def url_update(self, **kwargs):
        return parse.urlunparse(self._url._replace(**kwargs))

    def _request(self, typ, id=0, method="GET", params=None, data=None, url=None):
        """
        send the request, return response obj
        """

        headers = {"Accept": "application/json"}
        auth = None

        if self.user:
            auth = (self.user, self.password)

        if not url:
            if id:
                url = f"{self.url}/{typ}/{id}"
            else:
                url = f"{self.url}/{typ}"

        return requests.request(
            method, url, params=params, data=data, auth=auth, headers=headers
        )

    def _throw(self, res, data):
        self.log("=====> %s" % data)
        err = data.get("meta", {}).get("error", "Unknown")
        if res.status_code < 600:
            if res.status_code == 404:
                raise NotFoundException("%d %s" % (res.status_code, err))
            elif res.status_code == 401 or res.status_code == 403:
                raise PermissionDeniedException("%d %s" % (res.status_code, err))
            elif res.status_code == 400:
                raise InvalidRequestException("%d %s" % (res.status_code, err), data)

        # Internal
        raise Exception("%d Internal error: %s" % (res.status_code, err))

    def _load(self, res):
        try:
            data = res.json()
        except ValueError:
            data = {}

        if res.status_code < 300:
            if not data:
                return []
            return data.get("data", [])

        self._throw(res, data)

    def _mangle_data(self, data):
        if not "id" in data and "pk" in data:
            data["id"] = data["pk"]
        if "_rev" in data:
            del data["_rev"]
        if "pk" in data:
            del data["pk"]
        if "_id" in data:
            del data["_id"]

    def log(self, msg):
        if self.verbose:
            print(msg)

    def all(self, typ, **kwargs):
        """
        List all of type
        Valid arguments:
            skip : number of records to skip
            limit : number of records to limit request to
        """
        return self._load(self._request(typ, params=kwargs))

    def get(self, typ, id, **kwargs):
        """
        Load type by id
        """
        return self._load(self._request(typ, id=id, params=kwargs))

    def create(self, typ, data, return_response=False):
        """
        Create new type
        Valid arguments:
            skip : number of records to skip
            limit : number of records to limit request to
        """
        res = self._request(typ, method="POST", data=data)
        if res.status_code != 201:
            try:
                data = res.json()
                self._throw(res, data)
            except ValueError as e:
                if not isinstance(e, InvalidRequestException):
                    self._throw(res, {})
                else:
                    raise

        loc = res.headers.get("location", None)
        if loc and loc.startswith("/"):
            return self._load(self._request(None, url=self.url_update(path=loc)))
        if return_response:
            return res.json()

        return self._load(self._request(None, url=loc))

    def update(self, typ, id, **kwargs):
        """
        update just fields sent by keyword args
        """
        return self._load(self._request(typ, id=id, method="PUT", data=kwargs))

    def save(self, typ, data):
        """
        Save the dataset pointed to by data (create or update)
        """
        if "id" in data:
            return self._load(
                self._request(typ, id=data["id"], method="PUT", data=data)
            )

        return self.create(typ, data)

    def rm(self, typ, id):
        """
        remove typ by id
        """
        return self._load(self._request(typ, id=id, method="DELETE"))

    def type_wrap(self, typ):
        return TypeWrap(self, typ)


class TypeWrap:
    def __init__(self, client, typ):
        self.client = client
        self.typ = typ

    def all(self, **kwargs):
        """
        List all
        """
        return self.client.all(self.typ, **kwargs)

    def get(self, id):
        """
        Load by id
        """
        return self.client.get(self.typ, id)

    def save(self, data):
        """
        Save object
        """
        return self.client.save(self.typ, data)

    def rm(self, id):
        """
        remove by id
        """
        return self.client.rm(self.typ, id)
